### PR TITLE
MkdirTask behaves the same as "mkdir" Linux command and respects POSIX ACL

### PR DESCRIPTION
**Pull request by sustmi**
This PR makes `MkdirTask` create directories with the same permissions as Linux `mkdir` command. Also, it sets correct permissions when using [POSIX Access Control Lists](https://linux.die.net/man/5/acl).

It is possibly a BC break, but I think that it is worth it as the new behaviour is more intuitive.
### 1. a) Linux shell (ACL enabled):

```
$ setfacl --default --modify user::rwX .
$ getfacl .
# file: .
# owner: me
# group: me
user::rwx
group::rwx
other::r-x
default:user::rwx
default:user:me:rwx
default:group::rwx
default:mask::rwx
default:other::r-x

$ umask 0077
$ mkdir dir1
$ mkdir --mode 0777 dir2

$ ls -l
drwxrwxr-x+ 2 me me 4096 Oct 10 23:41 dir1
drwxrwxrwx+ 2 me me 4096 Oct 10 23:41 dir2

$ getfacl dir1
# file: dir1
# owner: me
# group: me
user::rwx
user:me:rwx
group::rwx
mask::rwx
other::r-x
default:user::rwx
default:user:me:rwx
default:group::rwx
default:mask::rwx
default:other::r-x
```

Notice that umask is completely ignored when ACL is enabled. Directory `dir1` has `rwxrwxr-x` permissions because that is the default setting of the parent directory.
#### build.xml:

``` ant
...
  <mkdir dir="dir1" />
  <mkdir dir="dir2" mode="0777" />
...
```
#### shell:

```
$ umask 0077
$ php phing

$ ls -l
drwx------+ 2 me me 4096 Oct 10 23.57 dir1
drwxrwxr-x+ 2 me me 4096 Oct 10 23.57 dir2

$ getfacl dir1
# file: dir1
# owner: me
# group: me
user::rwx
user:me:rwx         #effective:---
group::rwx          #effective:---
mask::---
other::---
default:user::rwx
default:user:me:rwx
default:group::rwx
default:mask::rwx
default:other::r-x
```

Notice that `dir1` has wrong permissions. Old `MkdirTask` calls `mkdir('dir1', 0700)` because it masks 0777 with the original umask (see the original `MkdirTask::__construct()`).
Also, you can see that in this case old `MkdirTask` incorrectly sets ACL mask to `---` because ACL mask equals to group permissions when ACL is enabled.
#### build.xml:

``` ant
...
  <mkdir dir="dir1" />
  <mkdir dir="dir2" mode="0777" />
...
```
#### shell:

```
$ umask 0077
$ php phing

$ ls -l
drwxrwxr-x+ 2 me me 4096 Oct 11 00.01 dir1
drwxrwxrwx+ 2 me me 4096 Oct 11 00.01 dir2

$ getfacl dir1
# file: dir1
# owner: me
# group: me
user::rwx
user:me:rwx
group::rwx
mask::rwx
other::r-x
default:user::rwx
default:user:me:rwx
default:group::rwx
default:mask::rwx
default:other::r-x
```

You can see that the new `MkdirTask` creates all permissions the same as Linux `mkdir` command. 
### 2. a) Linux shell (no ACL):

```
$ umask 0077
$ mkdir dir1
$ mkdir --mode 0777 dir2
$ mkdir --parents --mode 0777 dir3/dir4

$ ls -l
drwx------.   2 me me  4096 Oct 10 22:50 dir1
drwxrwxrwx.   2 me me  4096 Oct 10 22:51 dir2
drwx------.   2 me me  4096 Oct 10 22:51 dir3

$ ls -l dir3
drwxrwxrwx.   2 me me  4096 Oct 10 22:51 dir4
```

Notice that `dir2` is created with `0777` permissions even though umask is set to `0077`. That is because umask is ignored when `--mode` option is used.
Also notice that `dir3` has default permissions and not `0777` as specified in `--mode` option.
#### build.xml:

``` ant
...
  <mkdir dir="dir1" />
  <mkdir dir="dir2" mode="0777" />
  <mkdir dir="dir3/dir4" mode="0777" />
...
```
#### shell:

```
$ umask 0077
$ php phing

$ ls -l
drwx------. 2 me me 4096 Oct 10 23:32 dir1
drwxrwxrwx. 2 me me 4096 Oct 10 23:32 dir2
drwxrwxrwx. 3 me me 4096 Oct 10 23:32 dir3

$ ls -l dir3
drwxrwxrwx. 2 me me 4096 Ocr 10 23.32 dir4
```

Notice that `dir3` is not created with default permissions but with those specified in `mode` attribute which is not the same behaviour as in case of Linux `mkdir` command.
#### build.xml:

``` ant
...
  <mkdir dir="dir1" />
  <mkdir dir="dir2" mode="0777" />
  <mkdir dir="dir3/dir4" mode="0777" />
...
```
#### shell:

```
$ umask 0077
$ php phing

$ ls -l
drwx------. 2 me me 4096 Oct 10 23:36 dir1
drwxrwxrwx. 2 me me 4096 Oct 10 23:36 dir2
drwx------. 3 me me 4096 Oct 10 23:36 dir3

$ ls -l dir3
drwxrwxrwx. 2 me me 4096 Oct 10 23:36 dir4
```

`MkdirTask` now behaves the same as Linux `mkdir` command (ie. only last child directory has the permissions specified in `mode` attribute).

### CODE REVIEW

**Comment by sustmi** [comment]
I might rewrite the test-cases as proper PHPUnit tests if you agree with the changes in the PR.

**Comment by mrook** [comment]
I like the idea, but this is indeed a BC break - which makes it a nice candidate for the 3.0 release. So, could you:

* rebase this PR to the 3.0 branch
* add / rewrite test cases
* update the documentation where necessary

Thanks!

**Comment by sustmi** [comment]
Hi @mrook . I did not have much time to write the tests lately but I would like to finish the PR now.
Which branch should I rebase it to `unstable-3.0` or `future-4.0`?

**Comment by mrook** [comment]
@sustmi that would be great, please rebase against `master`.

**Comment by sustmi** [comment]
@mrook: The tests were added and the PR is rebased on `master`.
The tests for POSIX ACL are skipped on Travis CI. It probably does not support it (does not have `setfacl` command).
Also, I had to add one extra commit that fixes `ChmodTaskTest` to properly clean-up its testing `tmp` directory.
Should the documentation be updated too?

**Comment by sustmi** [comment]
Ok, I managed to enable ACL support in `.travis.yml` by installing `acl` system package so [the ACL tests are no longer skipped](https://travis-ci.org/phingofficial/phing/jobs/222587679#L783).

**Comment by mrook** [comment]
@siad007 could you review this one?

**Comment by mrook** [comment]
Yeah you're probably right. @sustmi can you make that happen?

**Comment by siad007** [comment]
I reopen this ticket. Related to #1088 - but we should think of windows support here.

**Comment by sustmi** [comment]
Sorry for leaving this PR stale. Are you still interested in getting this merged?
I changed a job and in the current one we do not use Phing so I do not really need this any more. But if there is somebody who will profit from this, we can try to get it done.

@siad007 : Can you be more concrete on what you think that should be documented?
I think I cannot "add **the whole** acl stuff to the documentation". :wink:

**Comment by siad007** [comment]
@sustmi I think we should only document the new behavior of the MkdirTask, so that all users are aware of this change.

File: test/classes/phing/tasks/system/MkdirTaskTest.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
+<?php
+
+require_once 'phing/BuildFileTest.php';
```
Comment (siad007) [inline test/classes/phing/tasks/system/MkdirTaskTest.php]: Please remove the `require_once` statement and add a license file comment.

Comment (sustmi) [reply test/classes/phing/tasks/system/MkdirTaskTest.php]: Done

**Comment by sustmi** [comment]
I rebased the PR on current master and fixed `MkdirTaskTest.php`.

> I think we should only document the new behavior of the MkdirTask, so that all users are aware of this change.

I was thinking about documenting the behavior in [MkdirTask documentation](https://www.phing.info/phing/guide/en/output/chunkhtml/MkdirTask.html) but it does not feel right because everything I can come up with is "it works like it should".

If the point is that users should notify the new behavior shouldn't be this just mentioned in `CHANGELOG.md`?
And since @mrook confirmed that this is a BC break, should I document it in `UPGRADING.md` too?
Or do you think that I should also add some "BC break in v XYZ" warning to the `MkdirTask` section?

**Comment by siad007** [comment]
> I was thinking about documenting the behavior in MkdirTask documentation but it does not feel right because everything I can come up with is "it works like it should".

> If the point is that users should notice the new behavior shouldn't be this just mentioned in CHANGELOG.md?
And since @mrook confirmed that this is a BC break, should I document it in UPGRADING.md too?
Or do you think that I should also add some "BC break in v XYZ" warning to the MkdirTask section?

As I dig deeper into this PR I feel the same about documentation and it would be the right way to add the changes to `CHANGELOG.md` and `UPGRADING.md`. There is indeed no need for the MkdirTask doc. But I noticed some other dependencies in other tasks, which relay on the default settings of the `PhingFile::mkdirs()` method, which could break existing behavior.

So please notice other tasks behavior changes:
```
Method
    mkdirs
Found usages  (15 usages found)
    Method call  (14 usages found)
        phingPhpArrayRebase  (14 usages found)
            classes\phing\system\io  (1 usage found)
                PhingFile.php  (1 usage found)
                    PhingFile  (1 usage found)
                        createNewFile  (1 usage found)
                            670 $parent->mkdirs();
            classes\phing\tasks\ext  (3 usages found)
                SmartyTask.php  (3 usages found)
                    SmartyTask  (3 usages found)
                        main  (2 usages found)
                            561 if (!$smartyCompilePath->mkdirs()) {
                            571 $file->mkdirs();
                        setOutputDirectory  (1 usage found)
                            261 if (!$outputDirectory->mkdirs()) {
            classes\phing\tasks\ext\phploc  (2 usages found)
                AbstractPHPLocFormatter.php  (1 usage found)
                    AbstractPHPLocFormatter  (1 usage found)
                        setToDir  (1 usage found)
                            77 $toDir->mkdirs();
                PHPLocTask.php  (1 usage found)
                    PHPLocTask  (1 usage found)
                        validateProperties  (1 usage found)
                            251 $reportOutputDir->mkdirs();
            classes\phing\tasks\ext\phpunit  (1 usage found)
                FormatterElement.php  (1 usage found)
                    FormatterElement  (1 usage found)
                        setToDir  (1 usage found)
                            105 $toDir->mkdirs();
            classes\phing\tasks\system  (4 usages found)
                CopyTask.php  (1 usage found)
                    CopyTask  (1 usage found)
                        doWork  (1 usage found)
                            539 if (!$d->mkdirs($dirMode)) {
                MkdirTask.php  (1 usage found)
                    MkdirTask  (1 usage found)
                        main  (1 usage found)
                            68 $result = $this->dir->mkdirs($this->mode);
                MoveTask.php  (1 usage found)
                    MoveTask  (1 usage found)
                        doWork  (1 usage found)
                            137 if (!$d->mkdirs()) {
                TruncateTask.php  (1 usage found)
                    TruncateTask  (1 usage found)
                        shouldProcess  (1 usage found)
                            140 $parent->mkdirs();
            classes\phing\util  (3 usages found)
                ExtendedFileStream.php  (1 usage found)
                    ExtendedFileStream  (1 usage found)
                        createDirectories  (1 usage found)
                            49 $f->mkdirs();
                FileUtils.php  (2 usages found)
                    FileUtils  (2 usages found)
                        copyFile  (1 usage found)
                            114 $parent->mkdirs($dirMode);
                        renameFile  (1 usage found)
                            170 $parent->mkdirs();
    Recursion  (1 usage found)
        phingPhpArrayRebase  (1 usage found)
            classes\phing\system\io  (1 usage found)
                PhingFile.php  (1 usage found)
                    PhingFile  (1 usage found)
                        mkdirs  (1 usage found)
                            769 return (($parentFile !== null) && ($parentFile->mkdirs($mode) && $this->mkdir($mode)));
```
I think for those we should explicity set the permissions to `0755` to avoid side effects.

**Comment by sustmi** [comment]
I added documentation to `UPGRADING.md`.
Should I created a new "Phing 3.0.0 alpha 4" or "Unreleased" section in `CHANGELOG.md`?

**Comment by siad007** [comment]
@sustmi the changes to the `UPGRADE.md` seems enough as the changelog should be auto generated.

**Comment by sustmi** [comment]
> So please notice other tasks behavior changes:

Wow, I did not realize that so much code depends on `mkdirs()` method. However, is there any use case where `mkdirs()` should not behave the same as `mkdir` Linux command?
I do not think that hard-coding 755 permissions is a good idea because I would expect the permissions of the newly created directories to obey Linux/POSIX standards.

**Comment by siad007** [comment]
Could you have a look [here](https://ci.appveyor.com/project/mrook/phing/builds/27839056):
```
BUILD FAILED
C:\projects\phing\test\build.xml:81:76 Test FAILURE (testParentDirectoriesHaveDefaultPermissions in class MkdirTaskTest): Failed asserting that file mode of "C:\projects\phing\test/etc/tasks/system/tmp/a" is 700
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'700'
+'777'
```

**Comment by siad007** [comment]
At least the `SmartyTask` uses it for the compiled files but I am not sure, which permission they relay on.

**Comment by sustmi** [comment]
Do you have any idea why the `testParentDirectoriesHaveDefaultPermissions` passed on Travis CI?
https://travis-ci.org/phingofficial/phing/jobs/592753673#L928
Or did they, because I do not see the test case name mentioned in the output. However it says:
```
  [phpunit] Testsuite: MkdirTaskTest
  [phpunit] Tests run: 7, Warnings: 0, Failures: 0, Errors: 0, Incomplete: 0, Skipped: 0, Time elapsed: 1.52346 s
```

Oh, I see, Appveyor runs the tests on Windows. 
The documentation of [mkdir()](https://www.php.net/manual/en/function.mkdir.php#refsect1-function.mkdir-parameters) php function says that:
> mode is ignored on Windows.

I guess we need to emulate the `$mode` parameter on Windows so that it behaves the same on both Windows and Linux.

Also, I found out that `mkdir()` PHP function has `$recursive` parameter. Why does not `\PhingFile::mkdirs()` use it? 
I found this old comment https://www.php.net/manual/en/function.mkdir.php#96990 that warns about the behavior that `mkdir()` does not set the `$mode` permissions to the created parent directories.
Maybe `MkdirTask` was consciously designed to be "smart" and set the permissions of the parent directories. But as it is not documented anywhere, I assume it should behave as standard `mkdir`.

**Comment by sustmi** [comment]
Never mind. I just realized that Windows does not support `umask`. The tests relying on `umask` should be skipped on Windows.

**Comment by sustmi** [comment]
I updated the `MkdirTaskTest` to skip scenarios that require non-zero umask (since Windows has umask hard-coded to zero).

**Comment by sustmi** [comment]
```
  [phpunit] testParentDirectoriesHaveDefaultPermissions with data set #0 FAILED
  [phpunit] Failed asserting that file mode of "C:\projects\phing\test/etc/tasks/system/tmp/a/b" is 707
```
Well, it looks like Windows may not even support `chmod()`.

**Comment by sustmi** [comment]
Ok, the `chmod()` on Windows is really "crooked".
See this test script:
```php
<?php

function chmodTest($file, $mode) {
    chmod($file, $mode);
    
    clearstatcache();
    var_dump(sprintf('expected: %03o, actual: %03o', $mode, fileperms($file)));
}

$file = __DIR__ . '/test.txt';

touch($file);

chmodTest($file, 0000);
chmodTest($file, 0777);
chmodTest($file, 0707);
chmodTest($file, 0700);
chmodTest($file, 0070);
chmodTest($file, 0007);

/*
Output on Windows 10, NTFS patition, php-7.3.10-Win32-VC15-x64:

string(29) "expected: 000, actual: 100444"
string(29) "expected: 777, actual: 100666"
string(29) "expected: 707, actual: 100666"
string(29) "expected: 700, actual: 100666"
string(29) "expected: 070, actual: 100444"
string(29) "expected: 007, actual: 100444"
*/
```

**Comment by sustmi** [comment]
I guess all the current tests in `MkdirTaskTest` can be skipped on Windows because they depend on POSIX permissions and those really do not work in PHP on Windows.

**Comment by sustmi** [comment]
Hooray! Tests pass both on Linux and Windows.
Tomorrow I will try to look at the code that uses `mkdirs()` to see if the new behavior could break anything.

**Comment by siad007** [comment]
> 
> 
> Hooray! Tests pass both on Linux and Windows.
> Tomorrow I will try to look at the code that uses `mkdirs()` to see if the new behavior could break anything.

@sustmi did you find some breaking stuff?

**Comment by siad007** [comment]
@sustmi finaly merged - thx for your commits and patience 👍

**Comment by sustmi** [comment]
Thank you for finishing it.
Unfortunatelly, I did not have a time to work on this any more.

**Pull request merged**